### PR TITLE
Lists Week 7 Milestone

### PR DESCRIPTION
**Pull request by lim-rachel**

### CODE REVIEW

File: compiler.ts [hunk 1]
```diff
@@ -107,6 +107,34 @@ function codeGenStmt(stmt: Stmt<[Type, SourceLocation]>, env: GlobalEnv): Array<
...
+        `, // TODO: Code under here will be changed if we implement negative indexing
+        ...getIndex, 
+        `(i32.const 0)
+        (i32.lt_s)
+        (if
+          (then
+            ${getIndex}
+            (call $list_index_oob)
+          )
+          (else)
```
Comment (yalhessi) [inline compiler.ts]: You can just delete this else since it's doing nothing.

File: ir.ts [hunk 2]
```diff
@@ -18,6 +18,7 @@ export type Stmt<A> =
   | {  a?: A, tag: "pass" }
   | {  a?: A, tag: "ifjmp", cond: Value<A>, thn: string, els: string }
   | {  a?: A, tag: "jmp", lbl: string }
+  | {  a?: A, tag: "check-index", address: Value<A>, index: Value<A> }
```
Comment (yalhessi) [inline ir.ts]: Adding to IR should be avoided at all costs if possible. The point is to keep IR minimal.

**Review by yalhessi — CHANGES_REQUESTED** [review]
Great work!

I really enjoyed going through the progress you've been making. Lots of great work here!

I encourage you to consider different designs for checking the validity of an index. We would certainly be very hesitant to approve anything that changes IR (or to a lesser degree compiler). If you consider what you're trying to do here, you're translating an index access to an index check, and then performing the access. This can easily be done in `lower` where you append a call to the statements being generated for both `index` and `index-assign`. To be more clear, you can replace line 297 for example with a `{tag: "call"}` where you call a function that does the checks. This will require you to pull the logic you embedded in `compiler` into a seperate wasm function, and make sure that wasm function is visible to our program. I think that would be the cleanest way to go about what you're trying to do.

File: compiler.ts (hunk 1 above)
Comment (xChickens) [reply compiler.ts]: Done.

File: ir.ts (hunk 2 above)
Comment (xChickens) [reply ir.ts]: Removed in latest commit.

**Review by xChickens — APPROVED** [review]

**Review by xChickens — APPROVED** [review]

**Pull request merged**